### PR TITLE
Fix NaN handling in cube fitting

### PR DESCRIPTION
**Pull request by rosteen**
I split this out from #3190 so it can be backported to 3.10. Still need to add a test and move the changelog.

Edit: ready for review. Note that simply adding `filter_non_finite` to the `specutils.fit_lines` call like in the 1D case didn't seem to work, so I masked out the NaN locations instead. Why that didn't work might be worth further investigation in the future.

### CODE REVIEW

File: jdaviz/configs/default/plugins/model_fitting/initializers.py [hunk 1]
```diff
@@ -190,7 +190,7 @@ def initialize(self, instance, x, y):
 
         # width can be estimated by the weighted
         # 2nd moment of the X coordinate.
-        dx = x - np.mean(x)
+        dx = x - np.nanmean(x)
```
Comment (pllim) [inline jdaviz/configs/default/plugins/model_fitting/initializers.py:193]: But `np.nanmean` would still return NaN if `x` is all NaNs. Is this something we need to handle?

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py [hunk 2]
```diff
@@ -917,6 +917,12 @@ def _fit_model_to_cube(self, add_data):
 
         # Apply masks from selected spectral subset
         spec = self._apply_subset_masks(spec, self.spectral_subset)
+        # Also mask out NaNs for fitting. Simply adding filter_non_finite to the cube fit
+        # didn't work out of the box, so doing this for now.
+        if spec.mask is None:
```
Comment (pllim) [inline jdaviz/configs/default/plugins/model_fitting/model_fitting.py:922]: Is this new logic tested?

Comment (rosteen) [reply jdaviz/configs/default/plugins/model_fitting/model_fitting.py:922]: The `None` case is, I should probably add a case to test the combination of existing mask and Nan-masking.

File: jdaviz/configs/default/plugins/model_fitting/initializers.py (hunk 1 above)
Comment (pllim) [reply jdaviz/configs/default/plugins/model_fitting/initializers.py:193]: Same question for `y` also.

Comment (pllim) [reply jdaviz/configs/default/plugins/model_fitting/initializers.py:193]: But didn't you say multiprocessing would swallow the exception, causing confusion?

Comment (pllim) [reply jdaviz/configs/default/plugins/model_fitting/initializers.py:193]: I could imagine someone did a bad subset that encompass all NaNs in the cube and then try to fit it. It is *possible*?

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py (hunk 2 above)
Comment (rosteen) [reply jdaviz/configs/default/plugins/model_fitting/model_fitting.py:922]: I added a test that should cover the existing spec.mask as well (that mask would come from a spectral subset).

**Comment by rosteen** [comment]
I have no idea why tests are failing on 3.11 and not 3.10 or 3.12, I'll need to investigate. 

Also note that I was initially testing the `spec.mask` combined with NaN-masking case by loading a `Spectrum1D` with a `mask` attribute before realizing that that loading case is not supported at right now - I fixed the first couple bugs I ran into with that, but stopped before going to far down the rabbit hole.

**Comment by rosteen** [comment]
The tests are passing locally for me on python 3.11, I don't see any package version differences that should affect them...I restarted the test and am crossing my fingers.

**Comment by pllim** [comment]
Hmm still fails. Could it be floating point differences? Maybe use allclose instead of equality? Also did you run with `--remote-data --run-slow` locally?

**Comment by pllim** [comment]
Try change

```python
assert np.all(result.get_component("flux").data == 1)
```

to

```python
np.testing.assert_allclose(result.get_component("flux").data, 1)
```

❓

**Comment by rosteen** [comment]
> Try change
> 
> ```python
> assert np.all(result.get_component("flux").data == 1)
> ```
> 
> to
> 
> ```python
> np.testing.assert_allclose(result.get_component("flux").data, 1)
> ```
> 
> ❓

No, you can see they're all 0s (as was the case before I fixed it) in that test environment rather than all 1s. 

I just confirmed that it also fails on my local WSL Ubuntu install, while still passing with an env set up the same way on my Mac, so at least I can reproduce it...

File: jdaviz/configs/default/plugins/model_fitting/tests/test_fitting.py [hunk 3]
```diff
@@ -378,3 +379,38 @@ def test_incompatible_units(specviz_helper, spectrum1d):
...
+    sv = cubeviz_helper.app.get_viewer('spectrum-viewer')
+    sv.apply_roi(XRangeROI(0, 5))
+
+    mf = cubeviz_helper.plugins["Model Fitting"]
+    mf.cube_fit = True
+    mf.spectral_subset = 'Subset 1'
+    mf.create_model_component("Const1D")
+    mf.calculate_fit()
+    result = cubeviz_helper.app.data_collection['model']
+    print(cubeviz_helper.app.data_collection['model'].get_object(statistic=None).flux.max())
```
Comment (pllim) [inline jdaviz/configs/default/plugins/model_fitting/tests/test_fitting.py:415]:
```suggestion
```

**Comment by rosteen** [comment]
Locally the tests pass on Ubuntu if I wrap the fitting call to ignore the `AstropyUserWarning` about linear fitting, even though that's not what the logs were showing as the error 🤯

**Comment by pllim** [comment]
Looks like warning is not always emitted.

```
E       Failed: DID NOT WARN. No warnings of type (<class 'AstropyUserWarning'>,) were emitted.
E        Emitted warnings: [].
```

Maybe just a blind ignore instead of pytest.warn

```python
with warnings.catch_warnings():
    warnings.filterwarnings(...)
```

**Comment by rosteen** [comment]
> Maybe just a blind ignore instead of pytest.warn
> 
> ```python
> with warnings.catch_warnings():
>     warnings.filterwarnings(...)
> ```

Yup I  just realized that's what was in the test above these.

**Review by pllim — APPROVED** [review]
Assuming CI is green this time, code changes look reasonable. Thanks!

**Comment by rosteen** [comment]
Whew, looks like that did it. Passing locally too.

**Review by gibsongreen — APPROVED** [review]
I've been following all the comments, and all the new tests look good and give the result I'd expect in a notebook, great work finishing this one off!

**Comment by pllim** [comment]
@rosteen is this worth backporting manually? If not, please update the milestone. And we also need follow-up PR to move the change log. Thanks!

**Pull request merged**